File: 13_fibonacci_numbers/fib_frog.cpp
#include <vector>

using namespace std;

vector<int> fib_under(int n) {
  vector<int> rt;
  int a = 1;
  int b = 1;
  int c;
  while (a <= n) {
    rt.push_back(a);
    c = b;
    b = a;
    a = b + c;
  }
  return rt;
}

int solution(vector<int> &a) {
  a.push_back(1);
  int n = a.size();
  vector<int> fibs = fib_under(n);
  vector<int> steps(n, n + 2);
  for (int i = 0; i < n; i++) {
    if (a[i] == 0) {
      continue;
    }
    for (vector<int>::iterator it = fibs.begin(); it != fibs.end(); it++) {
      if (i - *it < -1) {
        break;
      }
      else if (i - *it == -1) {
        steps[i] = 1;
      }
      else if (steps[i - *it] != n + 2) {
        steps[i] = min(steps[i], steps[i - *it] + 1);
      }
    }
  }
  if (steps[n - 1] == n + 2) {
    return -1;
  }
  else {
    return steps[n - 1];
  }
}

File: 13_fibonacci_numbers/fib_frog.py
def fib_under(n):
  a = 1
  b = 1
  rt = []
  while a <= n:
    rt.append(a)
    a, b = a + b, a
  return rt

def solution(a):
  a.append(1)
  n = len(a)
  fibs = fib_under(n)
  steps = [n + 2] * n
  for i in range(n):
    if a[i] == 0:
      continue
    for j in fibs:
      if i - j < -1:
        break
      elif i - j == -1:
        steps[i] = 1
      elif steps[i - j] != n + 2:
        steps[i] = min(steps[i], steps[i - j] + 1)
  if steps[-1] == n + 2:
    return -1
  else:
    return steps[-1]
